File: Threema/SharedUI/CustomFlows/ShareController.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2014-2023 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "ShareController.h"
#import "AppDelegate.h"
#import "UIDefines.h"
#import "ContactGroupPickerViewController.h"
#import "BundleUtil.h"
#import "Old_FileMessageSender.h"
#import "UTIConverter.h"
#import <CocoaLumberjack/CocoaLumberjack.h>
#import "UserSettings.h"

#ifdef DEBUG
static const DDLogLevel ddLogLevel = DDLogLevelVerbose;
#else
static const DDLogLevel ddLogLevel = DDLogLevelWarning;
#endif

@interface ShareController () <ContactGroupPickerDelegate, ModalNavigationControllerDelegate, UploadProgressDelegate>

@property NSMutableDictionary *info;

@end

@implementation ShareController

- (void)startShare {
    _info = [NSMutableDictionary dictionaryWithObjectsAndKeys:
             [NSNumber numberWithBool:YES], kKeyForceCompose,
             nil];
    
    if (_text) {
        [_info setObject:_text forKey:kKeyText];
    }

    if (_image) {
        [_info setObject:_image forKey:kKeyImage];
    }

    /* do we already know the target contact? */
    if (_contact != nil) {
        [self shareWithContact];
    } else {
        ModalNavigationController *navigationController = [ContactGroupPickerViewController pickerFromStoryboardWithDelegate:self];
        ContactGroupPickerViewController *picker = (ContactGroupPickerViewController *)navigationController.topViewController;
        picker.enableMultiSelection = NO;
        picker.enableTextInput = NO;
        picker.submitOnSelect = YES;
                
        [[AppDelegate sharedAppDelegate].window.rootViewController presentViewController:navigationController animated:YES completion:nil];
    }
}

- (void)shareWithContact {
    [_info setObject:_contact forKey:kKeyContact];
    [[NSNotificationCenter defaultCenter] postNotificationName:kNotificationShowConversation object:nil userInfo:_info];
}

- (void)shareWithConversation:(Conversation *)conversation renderType:(NSNumber *)renderType sendAsFile:(BOOL)sendAsFile {
    
    if (conversation == nil) {
        return;
    }
    
    if(_url) {
        NSSet *conversations = [NSSet setWithObject:conversation];
        
        URLSenderItem *item;
        if (sendAsFile) {
            NSString *mimetype = [UTIConverter mimeTypeFromUTI:[UTIConverter utiForFileURL:_url]];
            item = [URLSenderItem itemWithUrl:_url type:mimetype renderType:renderType sendAsFile:true];
        } else {
            item = [URLSenderItemCreator getSenderItemFor:_url];
        }
        for (Conversation *conv in conversations) {
            if ([UserSettings sharedUserSettings].newChatViewActive) {
                BlobManagerObjcWrapper *manager = [[BlobManagerObjcWrapper alloc] init];
                [manager createMessageAndSyncBlobsFor:item in:conversation correlationID:nil webRequestID:nil];
            }
            else {
                Old_FileMessageSender *sender = [[Old_FileMessageSender alloc] init];
                [sender sendItem:item inConversation:conv];
                sender.uploadProgressDelegate = self;
            }
        }
    } else {
        DDLogError(@"No URL provided, can't share anything");
    }
    
    [_info setObject:conversation forKey:kKeyConversation];
    [[NSNotificationCenter defaultCenter] postNotificationName:kNotificationShowConversation object:nil userInfo:_info];
}

- (void)deleteInboxFile {
    if ([_url.path rangeOfString:@"Documents/Inbox"].location != NSNotFound) {
        [[NSFileManager defaultManager] removeItemAtURL:_url error:nil];
    }
}

#pragma mark - Contact picker delegate

- (void)contactPicker:(ContactGroupPickerViewController*)contactPicker didPickConversations:(NSSet *)conversations renderType:(NSNumber *)renderType sendAsFile:(BOOL)sendAsFile {   
    [contactPicker dismissViewControllerAnimated:YES completion:^{
        // Only one expected
        Conversation *conversation = conversations.anyObject;
        [self shareWithConversation:conversation renderType:renderType sendAsFile:sendAsFile];
    }];
}

- (void)contactPickerDidCancel:(ContactGroupPickerViewController*)contactPicker {
    [contactPicker dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - ModalNavigationControllerDelegate

- (void)didDismissModalNavigationController {
    ;//nop
}

#pragma mark - UploadProgressDelegate

- (void)blobMessageSender:(Old_BlobMessageSender *)blobMessageSender uploadFailedForMessage:(BaseMessage *)message error:(UploadError)error {
    NSString *errorTitle = [BundleUtil localizedStringForKey:@"error_sending_failed"];
    NSString *errorMessage = [Old_FileMessageSender messageForError:error];
    
    [UIAlertTemplate showAlertWithOwner:[[AppDelegate sharedAppDelegate] currentTopViewController] title:errorTitle message:errorMessage actionOk:nil];    
    [self deleteInboxFile];
}

- (void)blobMessageSender:(Old_BlobMessageSender *)blobMessageSender uploadSucceededForMessage:(BaseMessage *)message {
    [self deleteInboxFile];
}

- (BOOL)blobMessageSenderUploadShouldCancel:(Old_BlobMessageSender *)blobMessageSender {
    return NO;
}

- (void)blobMessageSender:(Old_BlobMessageSender *)blobMessageSender uploadProgress:(NSNumber *)progress forMessage:(BaseMessage *)message {
    ;//nop
}

@end

File: Threema/SharedUI/CustomFlows/ShareController.swift
import CocoaLumberjackSwift
import FileUtility
import Foundation
import ThreemaFramework
import ThreemaMacros

final class ShareController: NSObject {

    private var info: [String: Any]?

    var text: String?
    var image: UIImage?
    var url: URL?

    // If nil, a contact picker will be shown
    var contact: ContactEntity?

    func startShare() {
        info = [kKeyForceCompose: NSNumber(booleanLiteral: true)]

        if let text {
            info?[kKeyText] = text
        }

        if let image {
            info?[kKeyImage] = image
        }

        // Do we already know the target contact?
        if let contact {
            share(of: contact)
        }
        else {
            guard let viewController = AppDelegate.shared().window.rootViewController,
                  let navigationController = ContactGroupPickerViewController.pickerFromStoryboard(withDelegate: self),
                  let picker = navigationController.topViewController as? ContactGroupPickerViewController else {
                return
            }

            picker.enableMultiSelection = false
            picker.enableTextInput = false
            picker.submitOnSelect = true

            viewController.present(navigationController, animated: true)
        }
    }

    // MARK: - Private functions

    private func share(of contactEntity: ContactEntity) {
        info?[kKeyContact] = contactEntity

        NotificationCenter.default.post(
            name: Notification.Name(kNotificationShowConversation),
            object: nil,
            userInfo: info
        )
    }

    private func share(of conversationEntity: ConversationEntity, renderType: Int, sendAsFile: Bool) {
        func setAndShowConversation() {
            info?[kKeyConversation] = conversationEntity

            NotificationCenter.default.post(
                name: Notification.Name(kNotificationShowConversation),
                object: nil,
                userInfo: info
            )
        }
        
        guard let url else {
            DDLogError("No URL provided, can't share anything")
            setAndShowConversation()
            return
        }
        
        var item: URLSenderItem?
        
        if sendAsFile {
            let mimeType =
                if let utiString = UTIConverter.uti(forFileURL: url),
                let mimeType = UTIConverter.mimeType(fromUTI: utiString) {
                    mimeType
                }
                else {
                    "application/octet-stream"
                }
            
            item = URLSenderItem(
                url: url,
                type: mimeType,
                renderType: NSNumber(integerLiteral: renderType),
                sendAsFile: true
            )
        }
        else {
            item = URLSenderItemCreator.getSenderItem(for: url)
        }
        
        if let item {
            Task { @MainActor in
                do {
                    try await BusinessInjector.ui.messageSender.sendBlobMessage(
                        for: item,
                        in: conversationEntity.objectID,
                        correlationID: nil,
                        webRequestID: nil
                    )
                    try deleteTmpFile(of: url)
                }
                catch {
                    DDLogError("Error while send and delete tmp file: \(error)")
                    try? deleteTmpFile(of: url)
                }
            }
        }
        setAndShowConversation()
    }

    private func deleteTmpFile(of url: URL) throws {
        guard let range = url.path.range(of: "tmp/Dropped"), !range.isEmpty else {
            return
        }
        try FileUtility.shared.delete(at: url)
    }
}

// MARK: - ContactGroupPickerDelegate

extension ShareController: ContactGroupPickerDelegate {
    func contactPicker(
        _ contactPicker: ContactGroupPickerViewController?,
        didPickConversations conversations: Set<AnyHashable>?,
        renderType: NSNumber?,
        sendAsFile: Bool
    ) {
        guard let contactPicker, let conversations, let renderType else {
            return
        }

        contactPicker.dismiss(animated: true) {
            // Only one expected
            guard let conversationEntity = conversations.first as? ConversationEntity else {
                return
            }
            self.share(of: conversationEntity, renderType: renderType.intValue, sendAsFile: sendAsFile)
        }
    }
    
    func contactPickerDidCancel(_ contactPicker: ContactGroupPickerViewController?) {
        guard let contactPicker else {
            return
        }

        contactPicker.dismiss(animated: true)
    }
}

// MARK: - ModalNavigationControllerDelegate

extension ShareController: ModalNavigationControllerDelegate {
    func didDismissModalNavigationController() {
        // no-op
    }
}
